File: csharp/poker/Poker.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Poker
{
    public enum Suit
    {
        Spade,
        Club,
        Heart,
        Diamond,
    }

    public class Card
    {
        public Card(int rank, Suit suit)
        {
            Rank = rank;
            Suit = suit;
        }

        public Card(string card)
        {
            Rank = ParseRank(card.Substring(0, card.Length - 1));
            Suit = ParseSuit(card[card.Length - 1]);
        }

        public int Rank { get; }
        public Suit Suit { get; }

        private int ParseRank(string rank)
        {
            switch (rank)
            {
                case "J": return 11;
                case "Q": return 12;
                case "K": return 13;
                case "A": return 14;
                default: return int.Parse(rank);
            }
        }

        private Suit ParseSuit(char value)
        {
            switch (value)
            {
                case 'S': return Suit.Spade;
                case 'C': return Suit.Club;
                case 'H': return Suit.Heart;
                case 'D': return Suit.Diamond;
                default: throw new ArgumentException("Invalid Suit");
            }
        }
    }

    public enum HandRank
    {
        HighCard,
        Pair,
        TwoPairs,
        ThreeOfAKind,
        Straight,
        Flush,
        FullHouse,
        FourOfAKind,
        StraightFlush,
    }

    public class Hand : IComparable<Hand>
    {
        private readonly Card[] rankedCards;
        private readonly HandRank handRank;

        public Hand(string input)
        {
            Input = input;
            var cards = input.Split(" ").Select(ci => new Card(ci)).OrderBy(c => c.Rank).ToArray();
            (handRank, rankedCards) = GetRank(cards);
        }

        public string Input { get; }

        public int CompareTo(Hand other)
        {
            if (handRank > other.handRank)
            {
                return 1;
            }
            else if (handRank < other.handRank)
            {
                return -1;
            }

            for (var i = 0; i < rankedCards.Length; i++)
            {
                if (rankedCards[i].Rank > other.rankedCards[i].Rank)
                {
                    return 1;
                }
                else if (rankedCards[i].Rank < other.rankedCards[i].Rank)
                {
                    return -1;
                }
            }

            return 0;
        }

        private (HandRank, Card[]) GetRank(Card[] cards)
        {
            var cardsByCount = cards.GroupBy(c => c.Rank).Select(g => new { count = g.Count(), card = g.First() } ).OrderByDescending(c => c.count).ThenByDescending(c => c.card.Rank).ToList();
            var rankedCards = cardsByCount.Select(c => c.card).ToArray();
            if (cardsByCount[0].count == 4)
            {
                return (HandRank.FourOfAKind, rankedCards);
            }

            if (cardsByCount[0].count == 3 && cardsByCount[1].count == 2)
            {
                return (HandRank.FullHouse, rankedCards);
            }

            if (cardsByCount[0].count == 3)
            {
                return (HandRank.ThreeOfAKind, rankedCards);
            }

            if (cardsByCount[0].count == 2 && cardsByCount[1].count == 2)
            {
                return (HandRank.TwoPairs, rankedCards);
            }

            if (cardsByCount.Count == 4)
            {
                return (HandRank.Pair, rankedCards);
            }

            var isFlush = cards.GroupBy(c => c.Suit).Count() == 1;
            var isStraight = (rankedCards[0].Rank == 14 && rankedCards[1].Rank == 5) || (rankedCards[0].Rank - rankedCards[4].Rank == 4);
            if (isStraight && rankedCards[0].Rank == 14)
            {
                rankedCards[0] = new Card(1, rankedCards[0].Suit);
            }

            if (isStraight && isFlush)
            {
                return (HandRank.StraightFlush, rankedCards);
            }
            else if (isStraight)
            {
                return (HandRank.Straight, rankedCards);
            }
            else if (isFlush)
            {
                return (HandRank.Flush, rankedCards);
            }

            return (HandRank.HighCard, rankedCards);
        }
    }

    public static IEnumerable<string> BestHands(IEnumerable<string> hands)
    {
        var allHands = hands.Select(input => new Hand(input)).ToList();
        allHands.Sort((x, y) => x.CompareTo(y));
        var bestHand = allHands.First();
        var bestHands = new List<Hand>() { bestHand };
        for (var i = 1; i < allHands.Count; i++)
        {
            var result = allHands[i].CompareTo(bestHand);
            if (result > 0)
            {
                bestHands.Clear();
                bestHand = allHands[i];
                bestHands.Add(bestHand);
            }
            else if (result == 0)
            {
                bestHands.Add(allHands[i]);
            }
        }

        return bestHands.Select(h => h.Input).ToArray();
    }
}

File: go/poker/poker.go
package poker

import (
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type suit int

const (
	Diamond suit = iota
	Heart
	Club
	Spade
)

type card struct {
	rank int
	suit suit
}

func newCard(c string) (*card, error) {
	runes := []rune(c)
	rank, err := parseCardRank(string(runes[0 : len(runes)-1]))
	if err != nil {
		return nil, err
	}

	suit, err := parseSuit(runes[len(runes)-1])
	if err != nil {
		return nil, err
	}

	return &card{rank: rank, suit: suit}, nil
}

func parseCardRank(input string) (int, error) {
	var rank int
	switch input {
	case "J":
		rank = 11
	case "Q":
		rank = 12
	case "K":
		rank = 13
	case "A":
		rank = 14
	default:
		parsedValue, err := strconv.Atoi(input)
		if err != nil || parsedValue < 2 || parsedValue > 10 {
			return 0, fmt.Errorf("%v is an invalid card rank", input)
		}
		rank = parsedValue
	}

	return rank, nil
}

func parseSuit(input rune) (s suit, err error) {
	switch input {
	case '♢':
		s = Diamond
	case '♡':
		s = Heart
	case '♤':
		s = Club
	case '♧':
		s = Spade
	default:
		err = fmt.Errorf("%c is an invalid suit", input)
	}
	return
}

type handRank int

const (
	HighCard handRank = iota
	Pair
	TwoPairs
	ThreeOfAKind
	Straight
	Flush
	FullHouse
	FourOfAKind
	StraightFlush
)

type hand struct {
	input            string
	handRank         handRank
	orderedCardRanks []int
}

func newHand(input string) (*hand, error) {
	var cardInputs = strings.Split(input, " ")
	if len(cardInputs) < 5 {
		return nil, errors.New("too few cards")
	}

	if len(cardInputs) > 5 {
		return nil, errors.New("too many cards")
	}

	cards := make([]*card, 0)
	for _, ci := range cardInputs {
		c, err := newCard(ci)
		if err != nil {
			return nil, err
		}
		cards = append(cards, c)
	}

	sort.Slice(cards, func(i, j int) bool {
		return cards[i].rank < cards[j].rank
	})

	handRank, cardRanks := getHandAndCardRanks(cards)

	return &hand{input: input, handRank: handRank, orderedCardRanks: cardRanks}, nil
}

func getHandAndCardRanks(cards []*card) (handRank, []int) {
	counts, cardRanks := getCountsAndCardRanks(cards)

	if counts[0] == 4 {
		return FourOfAKind, cardRanks
	}

	if counts[0] == 3 && counts[1] == 2 {
		return FullHouse, cardRanks
	}

	if counts[0] == 3 {
		return ThreeOfAKind, cardRanks
	}

	if counts[0] == 2 && counts[1] == 2 {
		return TwoPairs, cardRanks
	}

	if len(counts) == 4 {
		return Pair, cardRanks
	}

	var isFlush = true
	suit := cards[0].suit
	for i := 1; i < len(cards); i++ {
		if cards[i].suit != suit {
			isFlush = false
			break
		}
	}
	var isStraight = (cardRanks[0]-cardRanks[4] == 4) || (cardRanks[0] == 14 && cardRanks[1] == 5)
	if isStraight && cardRanks[0] == 14 {
		cardRanks[0] = 1
	}

	if isStraight && isFlush {
		return StraightFlush, cardRanks
	}

	if isStraight {
		return Straight, cardRanks
	}

	if isFlush {
		return Flush, cardRanks
	}

	return HighCard, cardRanks
}

type rankAndCount struct {
	rank  int
	count int
}

func compareRankAndCounts(a, b rankAndCount) bool {
	if a.count == b.count {
		return a.rank > b.rank
	}

	return a.count > b.count
}

func getCountsAndCardRanks(cards []*card) ([]int, []int) {
	countsByRank := make(map[int]rankAndCount)
	for _, c := range cards {
		countByRank, ok := countsByRank[c.rank]
		if ok {
			countByRank.count++
			countsByRank[c.rank] = countByRank
		} else {
			countsByRank[c.rank] = rankAndCount{rank: c.rank, count: 1}
		}
	}

	rankAndCounts := make([]rankAndCount, 0)
	for _, v := range countsByRank {
		rankAndCounts = append(rankAndCounts, v)
	}

	sort.Slice(rankAndCounts, func(i, j int) bool {
		return compareRankAndCounts(rankAndCounts[i], rankAndCounts[j])
	})

	counts := make([]int, 0)
	cardRanks := make([]int, 0)
	for i := 0; i < len(rankAndCounts); i++ {
		counts = append(counts, rankAndCounts[i].count)
		cardRanks = append(cardRanks, rankAndCounts[i].rank)
	}

	return counts, cardRanks
}

func compareHands(a, b *hand) int {
	if a.handRank > b.handRank {
		return 1
	} else if a.handRank < b.handRank {
		return -1
	}

	for i := range a.orderedCardRanks {
		if a.orderedCardRanks[i] > b.orderedCardRanks[i] {
			return 1
		} else if a.orderedCardRanks[i] < b.orderedCardRanks[i] {
			return -1
		}
	}

	return 0
}

// BestHand given poker hands returns best hands
func BestHand(inputs []string) ([]string, error) {
	bestHands := make([]string, 0)
	allHands := make([]*hand, 0)
	for _, input := range inputs {
		h, err := newHand(input)
		if err != nil {
			return bestHands, err
		}
		allHands = append(allHands, h)
	}

	bestHand := allHands[0]
	bestHands = append(bestHands, bestHand.input)
	for i := 1; i < len(allHands); i++ {
		r := compareHands(allHands[i], bestHand)
		if r > 0 {
			bestHands = bestHands[:0]
			bestHand = allHands[i]
			bestHands = append(bestHands, bestHand.input)
		} else if r == 0 {
			bestHands = append(bestHands, allHands[i].input)
		}
	}

	return bestHands, nil
}
